File: Python/basicOperations.py
from pyspark.sql import SparkSession, functions


def main() -> None:
    spark_session = SparkSession\
        .builder\
        .master("local[8]")\
        .getOrCreate()

    logger = spark_session._jvm.org.apache.log4j
    logger.LogManager.getLogger("org").setLevel(logger.Level.WARN)

    data_frame = spark_session \
        .read \
        .options(header='true', inferschema='true') \
        .option("delimiter", ",") \
        .csv("data/simple.csv") \
        .persist()

    data_frame.printSchema()
    data_frame.show()

    # Print the data types of the data frame
    print("data types: " + str(data_frame.dtypes))

    # Describe the dataframe
    data_frame\
        .describe()\
        .show()

    # Explain the dataframe
    data_frame \
        .explain()

    # select the names
    data_frame.select("Name").show()

    # select the columns name and age
    data_frame.select("Name", "Age").show()

    # select columns name and age but adding 2 to age
    data_frame.select("Name", data_frame["Age"] + 2).show()

    # select rows having a name length > 4
    data_frame.select(functions.length("Name") > 4).show()

    # select names starting with L
    data_frame\
        .select("Name", data_frame["Name"].startswith("L"))\
        .withColumnRenamed("startswith(Name,L)", "L-Name")\
        .show()

    # add a new column name "senior"
    data_frame.withColumn("Senior", data_frame["Age"] > 50).show()

    # rename HasACar to Owner
    data_frame.withColumnRenamed("HasACar", "Owner").show()

    # remove a column
    data_frame.drop("BirthDate").show()

    # sort by age
    data_frame.sort("Age").show()
    data_frame.sort(data_frame["Age"].desc()).show()

    # Get a RDD
    rdd_from_dataframe = data_frame\
        .rdd\
        .cache()

    for i in rdd_from_dataframe.collect():
        print(i)

    # Sum all the weights (RDD)
    sum_of_weights = rdd_from_dataframe\
        .map(lambda row: row[2])\
        .reduce(lambda x, y: x + y)
    print("Sum of weights (RDDs): " + str(sum_of_weights))

    v = data_frame.select("Weight").groupBy().sum().collect()
    print(v)
    print(v[0][0])

    # Sum all the weights (dataframe)
    weights = data_frame \
        .select("Weight") \
        .groupBy() \
        .sum() \
        .collect()

    print(weights)
    print("Sum of weights (dataframe): " + str(weights[0][0]))

    data_frame.select(functions.sum(data_frame["Weight"])).show()
    data_frame.agg({"Weight": "sum"}).show()

    # Get the mean age (RDD)
    total_age = rdd_from_dataframe\
        .map(lambda row: row[1]) \
        .reduce(lambda x, y: x + y)

    mean_age = total_age / rdd_from_dataframe.count()

    print("Mean age (RDDs): " + str(mean_age))

    # Get the mean age (dataframe)
    data_frame.select(functions.avg(data_frame["Weight"])) \
        .withColumnRenamed("avg(Weight)", "Average") \
        .show()

    data_frame.agg({"Weight": "avg"}).show()

    # Write to a json file
    data_frame\
        .write\
        .save("output.json", format="json")

    # Write to a CSV file
    data_frame\
        .write\
        .format("csv")\
        .save("output.csv")

if __name__ == "__main__":
    main()

File: JavaM9/src/main/java/org/mbd/m9/BasicOperations.java
package org.mbd.m9;

import org.apache.log4j.Level;
import org.apache.log4j.Logger;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.ColumnName;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.functions;
import static org.apache.spark.sql.functions.*;
import org.apache.spark.storage.StorageLevel;
import javax.xml.crypto.Data;
import java.util.List;

public class BasicOperations {
    public static void main(String[] args) {

        Logger.getLogger("org").setLevel(Level.OFF);

        SparkSession sparkSession =
                SparkSession.builder()
                    .appName("Java Spark SQL basic example")
                    .master("local[8]")
                    .getOrCreate();

        Dataset<Row> dataFrame = sparkSession
                .read()
                .format("csv")
                .option("header", "true")
                .load("data/simple.csv");

        dataFrame.printSchema();
        dataFrame.show();

        System.out.println("Data types: " + dataFrame.dtypes());

        dataFrame.describe().show();
        dataFrame.explain();

        //Select the names
        dataFrame.select("Name").show();

        //Select columns name and age
        dataFrame.select(col("Name"), col("Age")).show();

        //Select columns name and age but adding 2 to age
        dataFrame.select(col("Name"), col("Age").plus(2)).show();

        //Select the rows having a name length > 4
        dataFrame.select(functions.length(col("Name")).gt(4)).show();

        //Select names starting with "L"
        dataFrame
                .select(col("Name"), col("Name").startsWith("L"))
                .withColumnRenamed("startsWith(Name, L)", "L-names")
                .show();

        //Select names starting with "L"
        dataFrame.select(col("Name"), col("Name").startsWith("L")).show();

        //Add a new column name "Senior" containig true if age > 50
        dataFrame.withColumn("Senior", col("Age").gt(50)).show();

        //Rename HasACar as Owner
        dataFrame.withColumnRenamed("HasACar", "Owner").show();

        //Remove a column
        dataFrame.drop("BirthDate").show();

        //Sort by age
        dataFrame.sort(col("Age").desc()).show();
        dataFrame.orderBy(col("Age").desc(), col("Weight")).show();

        //Get a RDD from a DataFrame
        JavaRDD<Row> rddFromDataFrame = dataFrame.javaRDD().persist(StorageLevel.MEMORY_AND_DISK());

        for (Row row : rddFromDataFrame.collect()) {
            System.out.println(row);
        }

        //sum all weights (RDDs)
        double sumOfWeights = rddFromDataFrame
                .map(row -> Double.valueOf((String) row.get(2)))
                .reduce((x,y) -> x+y);

        System.out.println("Sum of weights (RDD): " + sumOfWeights);

        // sum all weights (dataframes)
        dataFrame.select("Weight").groupBy().sum().show();
        dataFrame.select(functions.sum("Weight")).show();
        dataFrame.agg(sum("Weight")).show();

        System.out.println("Sum of weights (RDD): " +
                dataFrame.agg(sum("weight")).collectAsList().get(0).get(0));

        //Compute the mean age (RDDs)
        int sumOfAges = rddFromDataFrame
                .map(row -> Integer.valueOf((String) row.get(1)))
                .reduce((x,y) -> x+y);

        System.out.println("Mean age: " + sumOfAges/rddFromDataFrame.count());

        //Compute the mean age (dataframe)
        dataFrame
                .select(functions.avg("Age"))
                .withColumnRenamed("avg(Age)", "Average")
                .show();

        //write to a csv file
        dataFrame.write().csv("output.csv");

        //Write to a JSON file
        dataFrame.write().json("output.json");

        sparkSession.stop();
    }
}

